halo2_proofs/src/plonk/vanishing/prover.rs

```diff
@@ -47,10 +49,25 @@ impl<C: CurveAffine> Argument<C> {
         transcript: &mut T,
     ) -> Result<Committed<C>, Error> {
         // Sample a random polynomial of degree n - 1
-        let mut random_poly = domain.empty_coeff();
-        for coeff in random_poly.iter_mut() {
-            *coeff = C::Scalar::random(&mut rng);
-        }
+        let n_threads = std::thread::available_parallelism().unwrap().get();
+        let needed_scalars = (1usize << domain.k() as usize) / n_threads;
+
+        let thread_seeds: Vec<ChaCha20Rng> = (0..n_threads)
+            .into_iter()
+            .map(|_| {
+                let mut seed = [0u8; 32];
+                rng.fill_bytes(&mut seed);
+                ChaCha20Rng::from_seed(seed)
+            })
+            .collect();
+
+        let rand_vec: Vec<C::Scalar> = thread_seeds
+            .par_iter()
+            .flat_map(|rng| vec![C::Scalar::random(rng.to_owned()); needed_scalars])
```

Comment:
Does `vec!` clone create a new random number for each vector entry?

Reply:
You're right that this wasn't as correct as it could be. 
I switched to:
```rust
let rand_vec: Vec<C::Scalar> = thread_seeds
            .par_iter_mut()
            .flat_map(|mut rng| {
                (0..needed_scalars)
                    .into_iter()
                    .map(|_| C::Scalar::random(&mut rng))
                    .collect::<Vec<C::Scalar>>()
            })
            .collect();
```

Good catch @jonathanpwang